### PR TITLE
Cleanup MPI with module deinit

### PR DESCRIPTION
**Pull request by npadmana**
The MPI module previously used a module-level record to
handle cleanup of the MPI module. With the introduction of
module level deinits, this is no longer necessary, and so we
remove the record, and use a module deinit.

Note that the issue of CHPL_COMM_WORLD being deallocated too 
early has now been resolved upstream (thanks!).

Tested on Linux --
 
 CHPL_COMM=none 
    (modules/packages/mpi/spmd/hello-chapel)
 CHPL_COMM=gasnet CHPL_COMM_SUBSTRATE=mpi 
   (modules/packages/mpi/multilocale/hello-chapel)

This partially addresses #5751 (point 3).

### CODE REVIEW

File: modules/packages/MPI.chpl [hunk 1]
```diff
@@ -208,42 +208,37 @@ module MPI {
...
-    proc deinit() {
-      if freeChplComm {
-        if numLocales > 1 {
-          coforall loc in Locales do on loc {
+  var _doinit : bool = false;
+  var _freeChplComm : bool = false;
+  // Module level deinit
+  proc deinit() {
+    if _freeChplComm {
+      if numLocales > 1 {
```
Comment (ben-albrecht) [inline modules/packages/MPI.chpl]: While we're here, differentiating `numLocales > 1` vs. `numLocales == 1` shouldn't be necessary here. The `coforall` should cover both cases.

**Comment by ben-albrecht** [comment]
Changes look good. My comment was an additional request.  I just need to test it before merging:

- [ ] Test on linux64
- [ ] Test on Cray

File: modules/packages/MPI.chpl (hunk 1 above)
Comment (npadmana) [inline modules/packages/MPI.chpl]: Fixed.

**Comment by npadmana** [comment]
Re-tested the two configurations described in the PR.

**Pull request merged**